launchpod/angular-launchpod/backend/src/test/java/com/google/launchpod/FileUploadServletTest.java

```diff
@@ -0,0 +1,666 @@
+    EmbeddedEntity mp3 = new EmbeddedEntity();
+    mp3.setProperty(ID, entityId);
+    mp3.setProperty(MP3_LINK, makeMp3Link(entityId));
+    mp3.setProperty(EMAIL, email);
+    return mp3;
+  }
+
+  /**
+   * Creates a test MP3 link.
+   */
+  private String makeMp3Link(String entityId) {
+    String link = "https://storage.googleapis.com/" + BUCKET_NAME + "/" + entityId;
+    return link;
+  }
+
+  /**
+   * Asserts that doPost() takes in form inputs from client and successfully
+   * stores that information in a Datastore entity.
+   */
+  @Test
+  public void doPost_StoresCorrectFormInput() throws IOException {
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+
+    when(request.getParameter(PODCAST_TITLE)).thenReturn(TEST_PODCAST_TITLE);
+    when(request.getParameter(DESCRIPTION)).thenReturn(TEST_DESCRIPTION);
+    when(request.getParameter(LANGUAGE)).thenReturn(TEST_LANGUAGE);
+    when(request.getParameter(EMAIL)).thenReturn(TEST_EMAIL);
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    servlet.doPost(request, response);
+
+    assertEquals(1, ds.prepare(new Query(USER_FEED)).countEntities(withLimit(10)));
+  }
+
+  /**
+   * Asserts that doPost() takes in form inputs from client, successfully
+   * stores that information in a Datastore entity, and correctly creates and stores xml string.
+   */
+  @Test
+  public void doPost_StoresCorrectFormInput_StoresCorrectXmlString() throws IOException {
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+
+    when(request.getParameter(PODCAST_TITLE)).thenReturn(TEST_PODCAST_TITLE);
+    when(request.getParameter(DESCRIPTION)).thenReturn(TEST_DESCRIPTION);
+    when(request.getParameter(LANGUAGE)).thenReturn(TEST_LANGUAGE);
+    when(request.getParameter(EMAIL)).thenReturn(TEST_EMAIL);
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    servlet.doPost(request, response);
+
+    Query query = new Query(USER_FEED);
+    PreparedQuery preparedQuery = ds.prepare(query);
+    Entity desiredEntity = preparedQuery.asSingleEntity();
+
+    String id = KeyFactory.keyToString(desiredEntity.getKey());
+    RSS testRSSFeed = new RSS(TEST_PODCAST_TITLE, TEST_DESCRIPTION, TEST_LANGUAGE, TEST_EMAIL, makeMp3Link(id));
+
+    // Verify xml string generation
+    String testXmlString = RSS.toXmlString(testRSSFeed);
+    assertEquals(testXmlString, desiredEntity.getProperty(XML_STRING).toString());
+  }
+
+  /**
+   * Asserts that doPost() takes in form inputs from client and successfully
+   * stores that information + MP3 information as an embedded entity in Datastore.
+   */
+  @Test
+  public void doPost_StoresCorrectFormInput_StoresCorrectMp3Info() throws IOException {
```

Comment:
all of these tests about just storing the form input are duplicated from the other servlet, there should be common class between both servlets and performs this logic, so you don't have duplicate tests testing the same logic. no need to do that here, but when you're deduping the code, that would be helpful

Reply:
makes a lot of sense. I'll also do this separately